File: server/src/application/entities/participant.ts
import { Replace } from '@helpers/Replace';
import { Conversation } from './conversation';
import { User } from './user';

interface ParticipantProps {
  userId: User['_id'];
  conversationId: Conversation['_id'];
  createdAt: Date;
  removedAt: Date | null;
}

export class Participant {
  private props: ParticipantProps;

  private constructor(props: ParticipantProps) {
    this.props = props;
  }

  public get userId() {
    return this.props.userId;
  }

  public get conversationId() {
    return this.props.conversationId;
  }

  public get createdAt() {
    return this.props.createdAt;
  }

  public get removedAt() {
    return this.props.removedAt;
  }

  public remove() {
    this.props.removedAt = new Date();
  }

  static create(props: Replace<ParticipantProps, { createdAt?: Date; removedAt?: Date }>) {
    return new Participant({
      ...props,
      createdAt: props.createdAt || new Date(),
      removedAt: props.removedAt || null,
    });
  }
}

File: Proxymity-Chat-Service/src/ProxyMity.Domain/Entities/Participant.cs
namespace ProxyMity.Domain.Entities;

[Table("participant")]
public class Participant
{
    public required Ulid UserId { get; set; }

    public required Ulid ConversationId { get; set; }

    public required DateTime CreatedAt { get; set; }

    public DateTime? RemovedAt { get; set; }

    public DateTime? UpdatedAt { get; set; }

    public DateTime? ConversationPinnedAt { get; set; }

    #region Foreign Keys

    [ForeignKey(nameof(UserId))]
    public User User { get; set; }

    [ForeignKey(nameof(ConversationId))]
    public Conversation Conversation { get; set; }

    #endregion

    public static Participant Create(Ulid userId, Ulid conversationId)
    {
        return new Participant
        {
            UserId = userId,
            ConversationId = conversationId,
            CreatedAt = DateTime.UtcNow,
        };
    }
}
